### ISSUE TITLE
Refresh access token failed with status code: 401

### ISSUE THREAD

**Issue by ericchansen** [microsoft/botframework-solutions#1326]
## Project

**Name**: Virtual Assistant

**Language**: C#

## Description

My bot does not deploy according to the documentation. Testing with the bot emulator results in POST 400 directline.postActivity with the contents

```
{
  "error": {
    "code": "ServiceError",
    "message": "Refresh access token failed with status code: 401"
  }
}
```

## To Reproduce

1. Follow the documentation at https://github.com/microsoft/AI/blob/master/docs/tutorials/csharp/virtualassistant.md.
2. Manually create the application registration and password.
3. For the deployment step, use something like

    ```
    .\Deployment\Scripts\deploy.ps1 -name "goose-the-cat" -location "westus" -luisAuthoringKey "xxx" -luisAuthoringRegion "westus" -resourceGroup "goose-the-cat" -appId "xxx" -appPassword "xxx"
    ```

    because you manually created the application registration.
4. Finish following the instructions at https://github.com/microsoft/AI/blob/master/docs/tutorials/csharp/virtualassistant.md.
5. It won't work!

```
{
  "error": {
    "code": "ServiceError",
    "message": "Refresh access token failed with status code: 401"
  }
}
```

## Expected behavior

To be able to send messages to the bot.

## Screenshots

![image](https://user-images.githubusercontent.com/5395779/57423688-98178c00-71da-11e9-9ead-124d39dc4a17.png)

## Additional context

Followed the directions to a T. Really not sure what's going on. To further complicate the issue, I've followed the directions to a T earlier today, and it worked.

**Comment by darrenj** [comment]
Hi Eric,

Looks like you manually created an AppId/Password on the AD application. Did it prompt you to do this as this should be auto-created for you. If you did create it, can I confirm you performed the manual creation steps detailed in the docs (namely the  Accounts in any organizational directory)

https://github.com/microsoft/AI/blob/master/docs/tutorials/assistantandskilldeploymentsteps.md

If so, can you validate that your emulator instance has the same appid/secret as in appsettings.config?

**Comment by ericchansen** [comment]
- I have tried with the automatic creation and the manual creation. I will say that with the automatic creation, I don't see the app registration in my Azure directory.
- I confirm that I did indeed set it to accounts in any organizational directory.
- I can also validate that the app ID and secret are identical to that found in appsettings.config (which is also the same as what's found in the Azure Portal with the manual creation).

**Comment by darrenj** [comment]
OK - adding support as this appears to be another variation on MSA App/Emulator issue perhaps? To also double check can you confirm the template version you have installed. Some deployment script changes were made on Monday but suspect you have this already.

![image](https://user-images.githubusercontent.com/33519520/57427073-225df100-71d7-11e9-95b1-87a163900880.png)

Pointer to latest just FYI is here in case you have an older template: https://github.com/microsoft/AI/tree/master/templates/Virtual-Assistant-Template/csharp/Sample/VirtualAssistantSample/Deployment/Scripts
visual

**Comment by ericchansen** [comment]
Looks like I'm up to date with the template version.

![image](https://user-images.githubusercontent.com/5395779/57427316-111de000-71e9-11e9-8910-b2848e7c6997.png)

**Comment by ericchansen** [comment]
I just uninstalled all of Node.js and reinstalled (who knows what's going on---I can't tell because the builds are working and the error messages inside the bot emulator aren't very useful).

```
PS C:\Users\eric.a.hansen> npm install -g botdispatch ludown@1.2.0 luis-apis qnamaker luisgen@2.0.2 botskills
npm WARN deprecated hoek@5.0.4: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).

> luisgen@2.0.2 preinstall C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\luisgen
> node dotnetpreinstall.js

C:\Users\eric.a.hansen\AppData\Roaming\npm\dispatch -> C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\botdispatch\bin\dispatch.js
C:\Users\eric.a.hansen\AppData\Roaming\npm\botskills -> C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\botskills\lib\botskills.js
C:\Users\eric.a.hansen\AppData\Roaming\npm\ludown -> C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\ludown\bin\ludown
C:\Users\eric.a.hansen\AppData\Roaming\npm\luis -> C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\luis-apis\bin\luis.js
C:\Users\eric.a.hansen\AppData\Roaming\npm\qnamaker -> C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\qnamaker\bin\qnamaker.js

> luisgen@2.0.2 install C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\luisgen
> dotnet tool install -g LUISGen

Tool 'luisgen' is already installed.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! luisgen@2.0.2 install: `dotnet tool install -g LUISGen`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the luisgen@2.0.2 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\eric.a.hansen\AppData\Roaming\npm-cache\_logs\2019-05-09T04_54_55_939Z-debug.log
PS C:\Users\eric.a.hansen> npm list -g botdispatch ludown luis-apis qnamaker luisgen botskills
C:\Users\eric.a.hansen\AppData\Roaming\npm
+-- botskills@1.0.0
| `-- botdispatch@1.4.5
+-- ludown@1.2.0
+-- luis-apis@2.5.1
`-- qnamaker@1.3.0
```

That doesn't look good?

**Comment by ericchansen** [comment]
Attempting redeployment again from scratch after all this npm business. If this works, then might want to update docs to 1st run `dotnet tool uninstall LUISGen --global`.

```
PS C:\Users\eric.a.hansen> dotnet tool uninstall LUISGen --global
Tool 'luisgen' (version '2.0.2') was successfully uninstalled.
PS C:\Users\eric.a.hansen> npm install -g luisgen@2.0.2

> luisgen@2.1.0 uninstall C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\luisgen
> dotnet tool uninstall --tool-path tools LUISGen

Tool 'luisgen' (version '2.1.0') was successfully uninstalled.

> luisgen@2.0.2 preinstall C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\luisgen
> node dotnetpreinstall.js


> luisgen@2.0.2 install C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\luisgen
> dotnet tool install -g LUISGen

You can invoke the tool using the following command: luisgen
Tool 'luisgen' (version '2.1.0') was successfully installed.
+ luisgen@2.0.2
updated 1 package in 3.038s
PS C:\Users\eric.a.hansen> npm install -g botdispatch ludown@1.2.0 luis-apis qnamaker luisgen@2.0.2 botskills
npm WARN deprecated hoek@5.0.4: This version has been deprecated in accordance with the hapi support policy (hapi.im/support). Please upgrade to the latest version to get the best features, bug fixes, and security patches. If you are unable to upgrade at this time, paid support is available for older versions (hapi.im/commercial).

> luisgen@2.0.2 uninstall C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\luisgen
> dotnet tool uninstall -g LUISGen

Tool 'luisgen' (version '2.1.0') was successfully uninstalled.

> luisgen@2.0.2 preinstall C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\luisgen
> node dotnetpreinstall.js

C:\Users\eric.a.hansen\AppData\Roaming\npm\dispatch -> C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\botdispatch\bin\dispatch.js
C:\Users\eric.a.hansen\AppData\Roaming\npm\botskills -> C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\botskills\lib\botskills.js
C:\Users\eric.a.hansen\AppData\Roaming\npm\ludown -> C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\ludown\bin\ludown
C:\Users\eric.a.hansen\AppData\Roaming\npm\luis -> C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\luis-apis\bin\luis.js
C:\Users\eric.a.hansen\AppData\Roaming\npm\qnamaker -> C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\qnamaker\bin\qnamaker.js

> luisgen@2.0.2 install C:\Users\eric.a.hansen\AppData\Roaming\npm\node_modules\luisgen
> dotnet tool install -g LUISGen

You can invoke the tool using the following command: luisgen
Tool 'luisgen' (version '2.1.0') was successfully installed.
+ luisgen@2.0.2
+ ludown@1.2.0
+ botdispatch@1.4.5
+ luis-apis@2.5.1
+ qnamaker@1.3.0
+ botskills@1.0.0
updated 6 packages in 15.972s
PS C:\Users\eric.a.hansen> npm list -g botdispatch ludown luis-apis qnamaker luisgen botskills
C:\Users\eric.a.hansen\AppData\Roaming\npm
+-- botskills@1.0.0
| `-- botdispatch@1.4.5
+-- ludown@1.2.0
+-- luis-apis@2.5.1
+-- luisgen@2.0.2
`-- qnamaker@1.3.0
```

**Comment by ericchansen** [comment]
The npm idea did not fix my problem. I'm seeing the same behavior.

```
{
  "error": {
    "code": "ServiceError",
    "message": "Refresh access token failed with status code: 401"
  }
}
```

**Comment by mdrichardson** [comment]
@ericchansen CC: @darrenj 

I just followed all of the steps in the REAMDE and was unable to produce the 401 error. 

* I did notice, however, that the powershell command fails if the bot files are in a directory with a space in the name. Did your Powershell command complete successfully?

* 401 is almost always a mismatched appId/appPassword. You've stated that they match in your App Registration as well as `appsettings.json`. Can you also verify that they match in Azure > App Service > Configuration?

![image](https://user-images.githubusercontent.com/40401643/57476825-44458b00-724c-11e9-9239-41dfbfc4186e.png)

That being said, it looks like you're debugging locally, connecting to `localhost`. You don't need `appId` and `appPassword` for that. 

* Can you follow the [auth troubleshooting instruction here](https://docs.microsoft.com/en-us/azure/bot-service/bot-service-troubleshoot-authentication-problems?view=azure-bot-service-4.0) and let me know how it goes?
* Is your Emulator up-to-date?

*Note: You may need to fully close out and restart Emulator between tests when removing or changing `appId` and `appPassword`. The only way I could seem to get a 401 error would be to test w/ ID/pass, remove the ID/pass from appsettings.json, restart the bot, then try to test in Emulator without ID/pass, without restarting Emulator. Once I restarted Emulator, I could test without ID/pass just fine*

**Comment by ericchansen** [comment]
Looks like I'm on 4.3.3. What should I be on?

![image](https://user-images.githubusercontent.com/5395779/57479525-3d764280-7263-11e9-92e3-1c1916279c63.png)

I can click *Help > Restart to Update...* but that doesn't actually update the program. It also doesn't restart. It just closes the program and I have to reopen manually.

I'll run through your other suggestions now.

**Comment by ericchansen** [comment]
@mdrichardson I'm curious about a handful of things. When you followed the steps in the documentation, did you manually create the application registration? Did the name of the application registration have anything in common with the name of your bot solution? Was the application registration in the same Azure sub as the other bot resources? Also, if you run `npm list -g botdispatch ludown luis-apis qnamaker luisgen botskills`, what is the output?

**Comment by mdrichardson** [comment]
[4.4.1 is the latest](https://github.com/microsoft/BotFramework-Emulator/releases). May as well update to rule that out. Sometimes the update just doesn't work. You either have to download from that link, or if it says "Restart to Update", you can go to `%AppData%\Roaming\@bfemulator\main\__update__` and run the `.exe` manually.

I'll answer your questions directed @darrenj, for now--he's likely busy with Microsoft Build 2019--Virtual Assistant is a big part of it.

I didn't do the manual creation, but I'll try that now.

I would think that the app needs to be under the same subscription as the other resources, but haven't tested. Will try that, too.

Here's my output of `npm list -g botdispatch ludown luis-apis qnamaker luisgen botskills`:

```bash
+-- botskills@1.0.0
| `-- botdispatch@1.4.5
+-- ludown@1.3.0
+-- luis-apis@2.5.1
+-- luisgen@2.1.0
`-- qnamaker@1.3.0
```

**Comment by ericchansen** [comment]
Apologies for linking the wrong person in my last comment!

Your npm package versions contradict what's in the documentation. You're on ludown 1.3.0 when it should be 1.2.0. Similarly, the docs say we should use luisgen 2.0.2, but you have 2.1.0. I wonder if the docs are wrong or outdated. It's funny though because @darrenj specified those version numbers in the docs yesterday 2873a84c704399215a0f3117533f9cbd3b2fb22b (excerpt below).

```
npm install -g botdispatch ludown@1.2.0 luis-apis qnamaker luisgen@2.0.2 botskills
```

Also, I just ran a deployment on a fresh project (output below). This looks exactly as the docs intend.

```
.\Deployment\Scripts\deploy.ps1 -name "ThisIsAGenuinelyNewName" -location "westus" -luisAuthoringKey "2499ed45abfe4291add663af2aca28ab" -luisAuthoringRegion "westus" -resourceGroup "ThisIsAGenuinelyNewName" -appId "93047bb4-ecbf-425f-bc61-ab82ad03aed5" -appPassword "?HvwJ-Pu2widYo9e6e?[IvcUxMdmzlb5"
> Creating resource group ...
> Deploying Azure services (this could take a while)...
> Updating appsettings.json ...
> Deploying cognitive models ...
> Initializing dispatch model ...
> Parsing general LU file ...

     Update available 1.2.0 -> 1.3.0
     Run npm i -g ludown to update.
> Deploying general LUIS app ...
> Adding general app to dispatch model ...
> Parsing chitchat LU file ...

     Update available 1.2.0 -> 1.3.0
     Run npm i -g ludown to update.
> Deploying chitchat QnA kb ...
> Adding chitchat kb to dispatch model ...
> Parsing faq LU file ...

     Update available 1.2.0 -> 1.3.0
     Run npm i -g ludown to update.
> Deploying faq QnA kb ...
> Adding faq kb to dispatch model ...
> Creating dispatch model...
> Done.
```

**Comment by ericchansen** [comment]
Well I'm extra confused now.

![image](https://user-images.githubusercontent.com/5395779/57480940-a1e6d100-7266-11e9-95f3-d87fa0c7568f.png)

This is with the npm package versions specified in the docs (see below) and Bot Emulator 4.3.3.

```
npm list -g botdispatch ludown luis-apis qnamaker luisgen botskills
C:\Users\eric.a.hansen\AppData\Roaming\npm
+-- botskills@1.0.0
| `-- botdispatch@1.4.5
+-- ludown@1.2.0
+-- luis-apis@2.5.1
+-- luisgen@2.0.2
`-- qnamaker@1.3.0
```

Is it possible that there could be transient issues? What possible causes could there be that would cause temporary outage?

Notice that in my [Stack Overflow thread on this same topic](https://stackoverflow.com/questions/56047387/deploying-and-running-bot-framework-v4-4-from-template-works-but-deploying-and) I commented on how the problem seemed to appear out of nowhere on a deployment that was previously functioning.

* Temporary Azure service outage last evening?
* Bot Emulator or some other piece might not tolerate certain appID or appPassword strings?
* That's all I can imagine for now.

**Comment by ericchansen** [comment]
Another idea. I can't gaurantee the other tests I did had 100% new names for the LUIS app, app registration, solution name, etc. I would delete all the resources, all of the solution, etc. and perhaps reuse the name in a test later on. However, this time I'm sure the name is 100% new. I wonder if the Bot Emulator would reuse an appId or appPassword that was cached somewhere, somehow? Hmm.

**Comment by ericchansen** [comment]
Confirmed that it still works with Bot Emulator 4.4.1.

**Comment by mdrichardson** [comment]
My guess regarding the npm packages is that the latest had some errors--I know there's been a couple issues related to the LUIS tools in the last couple of days. Shouldn't affect this error, though, unless it stops deployment-which it didn't.

Odd that it started working now. 

I haven't seen any reported outages in the last couple of days.

I've definitely had issues with some password strings. [The README](https://github.com/microsoft/AI/blob/master/docs/tutorials/assistantandskilldeploymentsteps.md#my-microsoft-app-registration-could-not-be-automatically-provisioned) notes that it has trouble with passwords containing `@`. I know I've had trouble with another password, however (I don't remember what special character gave it the issue). I would *guess* that this was the issue.

I've re-used names and not had issues before. Could be transient, I guess.

My best guess is that it was either an issue with a special character in a password, emulator caching id/pass in some unexpected way, or IIS Express caching id/pass in some way. Usually, if I'm switching bots with the same endpoints and running into trouble, I restart those and it usually works.

I'm going to close this as it appears to be resolved. If you run into this again or feel it isn't resolved, feel free to comment/re-open. And thank you for being willing to troubleshoot and providing so much detail!!

Edit: I tried with creating the App Registration manually and it worked fine, too, even if I specified an unmatching name.
